deeptools/heatmapper.py

```diff
@@ -647,33 +647,75 @@ def save_matrix(self, file_name):
-
-        # TODO this function must be updated
-        print "save tabulated values is not yet implemented."
+    def save_tabulated_values(self, file_handle, reference_point_label='TSS', start_label='TSS', end_label='TES', averagetype='mean'):
         """
-        avgDict = OrderedDict()
-        stdDict = OrderedDict()
+        Saves the values averaged by col using the avg_type
+        given
 
-        for label, heatmapMatrix in self.matrixDict.iteritems():
-            avgDict[label] = heatmapper.matrix_avg(heatmapMatrix, 'mean')
-            stdDict[label] = heatmapper.matrix_avg(heatmapMatrix, 'std')
+        Args:
+            file_handle: file name to save the file
+	    reference_point_label: Name of the reference point label
+	    start_label: Name of the star label
+	    end_label: Name of the end label
+            avg_type: average type (e.g. mean, median, std)
 
-        file_handle.write(
-            '#bin No.\t{}\n'.format(" mean\t std\t".join(avgDict.keys())))
+        """
+        #  get X labels
+        w = self.parameters['bin size']
+        b = self.parameters['upstream']
+        a = self.parameters['downstream']
+        m = self.parameters['body']
+
+        if b < 1e5:
+            quotient = 1000
+            symbol = 'Kb'
+        else:
+            quotient = 1e6
+            symbol = 'Mb'
 
-        for j in range(0, len(avgDict[avgDict.keys()[0]])):
-            file_handle.write('{}\t'.format(bin[j]))
-            for label in self.matrixDict.keys():
-                file_handle.write(
-                    '{}\t{}\t'.format(avgDict[label][j], stdDict[label][j]))
-            file_handle.write('\n')
+        if m == 0:
+            xticks = [(k / w)  for k in [w, b, b + a]]
+            xtickslabel = ['{0:.1f}{1}'.format(-(float(b) / quotient), symbol), reference_point_label,
+                           '{0:.1f}{1}'.format(float(a) / quotient, symbol)]
 
-        file_handle.close()
-        """
+        else:
+            xticks_values = [w]
+            xtickslabel = []
+
+            # only if upstream region is set, add a x tick
+            if self.parameters['upstream'] > 0:
+                xticks_values.append(b)
+                xtickslabel.append('{0:.1f}{1}'.format(-(float(b) / quotient), symbol))
+
+            # set the x tick for the body parameter, regardless if
+            # upstream is 0 (not set)
+            xticks_values.append(b + m)
+            xtickslabel.append(start_label)
+            xtickslabel.append(end_label)
+            if a > 0:
+                xticks_values.append(b + m + a)
+                xtickslabel.append('{0:.1f}{1}'.format(float(a) / quotient, symbol))
+
+            xticks = [(k / w)  for k in xticks_values]
+        x_axis = np.arange(xticks[-1]) + 1
+        labs = []
+        for x_value in x_axis:
+            if x_value in xticks:
+             labs.append(xtickslabel[xticks.index(x_value)])
+            else:
+             labs.append("")
+
```

Comment:
All this code above is simply to get some labels for the x-axis of the output that match what is plotted. Otherwise is just bins ids.